// graojs/tools/bundles/teste/TesteController.js
// Glocal Scope :)
var models, 
	controllers,
	event,
	teste, // object
	Teste; // object/class
var service = {
	get : function(req, res) {
			Teste.findOne({_id : req.params.id}, function(err, Teste) {
			if (err)
			{
				event.newEvent(err).error().present().log('error');
				res.jsonp(teste);
				res.end();
			}
		});
	},
	query : function(req, res) {
		Teste.find().sort('-created').populate('Teste').exec(function(err, testes) {
			if (err) {
				event.newEvent(err).error().present().log('error');
				res.end();
			} else {
				res.jsonp(testes);
				res.end();
			}
		});
	},
	create : function(req, res) {
		Teste = new Teste(req.body);
		teste.save(function(err) {
			if (err) {
				event.newEvent(err).error().present().log('error');
			} else {
				event.newEvent('created').success().present().log('info');
			}
		});
		res.end();
	},
	update : function(req, res) {
		delete req.body._id;
		Teste.findOneAndUpdate({_id : req.params.id }, req.body, { upsert : true }, function(err, Teste) {
			if (err) {
				event.newEvent(err).error().present().log('error');
			} else {
				try {
					event.newEvent('updated').success().present().log('info');
				} catch(err) {
					event.newEvent(err).error().present().log('error');
				}
			}
		});
		res.end();
	},
	destroy : function(req, res) {	
		Teste.remove({_id : req.params.id}, function(err) {
			if (err) {
				event.newEvent(err).error().present().log('error');
			} else {
				event.newEvent('destroyed').success().present().log('info');
			}
		});
		res.end();
	}
};
var admin = {
	dashboard : function(req, res) {
		res.render('teste/view/dashboard');
	}
};
var TesteController = function(di) {
	event = new di.event.newEvent('Instance created').success().present().log('info');
	models = di.models;
	controllers = di.controllers;
	Teste = models.Teste; // object/class
	Teste = new Teste(); // object
	this.service = service;
	this.admin = admin;
};
module.exports = exports = TesteController;
